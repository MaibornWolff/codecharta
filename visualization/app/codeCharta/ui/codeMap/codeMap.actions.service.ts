import { CodeMapNode, EdgeVisibility, MarkedPackage } from "../../codeCharta.model"
import { StoreService } from "../../state/store.service"
import { setEdges } from "../../state/store/fileSettings/edges/edges.actions"
import { unmarkPackage, setMarkedPackages } from "../../state/store/fileSettings/markedPackages/markedPackages.actions"
import { EdgeMetricDataService } from "../../state/store/metricData/edgeMetricData/edgeMetricData.service"
import { getParent } from "../../util/nodePathHelper"

export class CodeMapActionsService {
	constructor(private edgeMetricDataService: EdgeMetricDataService, private storeService: StoreService) {}

	markFolder({ path }: CodeMapNode, color: string) {
		const { markedPackages } = this.storeService.getState().fileSettings
		const markedPackagesMap = new Map(markedPackages.map(entry => [entry.path, entry]))

		const markedPackage = getParent(markedPackagesMap, path)

		if (!markedPackage || markedPackage.color !== color) {
			markedPackagesMap.set(path, {
				path,
				color
			})
		}

		for (const [key, markedPackage] of markedPackagesMap) {
			if (markedPackage.path === path) {
				if (markedPackage.color !== color) {
					markedPackagesMap.delete(key)
				}
			} else if (markedPackage.path.startsWith(path)) {
				// Remove marked packages with color identical to their parent marked package
				const markedPackageParent = getParent(markedPackagesMap, markedPackage.path)
				if (markedPackageParent && markedPackageParent.color === markedPackage.color) {
					markedPackagesMap.delete(key)
				}
			}
		}

		this.storeService.dispatch(setMarkedPackages([...markedPackagesMap.values()]))
	}

	unmarkFolder(node: CodeMapNode) {
		let index = this.storeService.getState().fileSettings.markedPackages.findIndex(mp => mp.path === node.path)

		if (index === -1) {
			index = this.getParentMarkedPackageIndex(node.path)
		}

		if (index !== -1) {
			this.storeService.dispatch(unmarkPackage(index))
		}
	}

	updateEdgePreviews() {
		const state = this.storeService.getState()
		const { edges } = state.fileSettings
		const { edgeMetric } = state.dynamicSettings
		const numberOfEdgesToDisplay = state.appSettings.amountOfEdgePreviews
		const edgePreviewNodes = new Set(this.edgeMetricDataService.getNodesWithHighestValue(edgeMetric, numberOfEdgesToDisplay))

		for (const edge of edges) {
			edge.visible = EdgeVisibility.none

			if (edge.attributes[edgeMetric] !== undefined) {
				const hasFromNodeEdgePreview = edgePreviewNodes.has(edge.fromNodeName)
				const hasToNodeEdgePreview = edgePreviewNodes.has(edge.toNodeName)

				if (hasFromNodeEdgePreview === hasToNodeEdgePreview) {
					if (hasFromNodeEdgePreview) {
						edge.visible = EdgeVisibility.both
					}
				} else if (hasFromNodeEdgePreview) {
					edge.visible = EdgeVisibility.from
				} else {
					edge.visible = EdgeVisibility.to
				}
			}
		}

		this.storeService.dispatch(setEdges(edges))
	}

	getParentMarkedPackageIndex(path: string) {
		const markedPackages: MarkedPackage[] = this.storeService.getState().fileSettings.markedPackages
		let index = -1
		for (let loopIndex = 0; loopIndex < markedPackages.length; loopIndex++) {
			const markedPackage = markedPackages[loopIndex]
			if (
				path.startsWith(markedPackage.path) &&
				path !== markedPackage.path &&
				(index === -1 || markedPackages[index].path.length < markedPackage.path.length)
			) {
				index = loopIndex
			}
		}

		return index
	}
}
