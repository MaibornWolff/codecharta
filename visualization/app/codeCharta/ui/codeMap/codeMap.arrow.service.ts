import { BuildingDeselectedEventSubscriber, BuildingSelectedEventSubscriber, ThreeSceneService } from "./threeViewer/threeSceneService"
import { Node, EdgeVisibility, Edge } from "../../codeCharta.model"
import { ArrowHelper, BufferGeometry, CubicBezierCurve3, Line, LineBasicMaterial, Object3D, Vector3 } from "three"
import { BuildingHoveredSubscriber, CodeMapMouseEventService, BuildingUnhoveredSubscriber } from "./codeMap.mouseEvent.service"
import { IRootScopeService } from "angular"
import { ColorConverter } from "../../util/color/colorConverter"
import { CodeMapBuilding } from "./rendering/codeMapBuilding"
import { StoreService } from "../../state/store.service"

export class CodeMapArrowService
	implements BuildingSelectedEventSubscriber, BuildingDeselectedEventSubscriber, BuildingHoveredSubscriber, BuildingUnhoveredSubscriber {
	private VERTICES_PER_LINE = 5
	private map: Map<string, Node>
	private arrows: Object3D[]

	constructor(private $rootScope: IRootScopeService, private storeService: StoreService, private threeSceneService: ThreeSceneService) {
		this.arrows = new Array<Object3D>()
		CodeMapMouseEventService.subscribeToBuildingHovered(this.$rootScope, this)
		CodeMapMouseEventService.subscribeToBuildingUnhovered(this.$rootScope, this)
		ThreeSceneService.subscribeToBuildingSelectedEvents(this.$rootScope, this)
		ThreeSceneService.subscribeToBuildingDeselectedEvents(this.$rootScope, this)
	}

	onBuildingSelected(selectedBuilding: CodeMapBuilding) {
		if (this.isEdgeApplicableForBuilding(selectedBuilding)) {
			this.clearArrows()
			this.showEdgesOfBuildings()
		}
		this.scale()
	}

	onBuildingDeselected() {
		this.clearArrows()
		this.addEdgePreview(
			null,
			this.storeService.getState().fileSettings.edges.filter(x => x.visible !== EdgeVisibility.none)
		)
	}

	onBuildingHovered(hoveredBuilding: CodeMapBuilding) {
		if (this.isEdgeApplicableForBuilding(hoveredBuilding)) {
			this.clearArrows()
			this.showEdgesOfBuildings(hoveredBuilding)
		}
		this.scale()
	}

	onBuildingUnhovered() {
		const state = this.storeService.getState()
		if (state.dynamicSettings.edgeMetric !== "None") {
			this.clearArrows()
			this.showEdgesOfBuildings()
		}
		this.scale()
	}

	clearArrows() {
		this.arrows = []
		while (this.threeSceneService.edgeArrows.children.length > 0) {
			this.threeSceneService.edgeArrows.children.pop()
		}
	}

	addArrow(arrowTargetNode: Node, arrowOriginNode: Node, buildingIsOriginNode: boolean) {
		const state = this.storeService.getState()
		const curveScale = 100 * state.appSettings.edgeHeight

		if (
			arrowTargetNode.attributes?.[state.dynamicSettings.heightMetric] &&
			arrowOriginNode.attributes?.[state.dynamicSettings.heightMetric]
		) {
			const curve = this.createCurve(arrowOriginNode, arrowTargetNode, curveScale)

			if (buildingIsOriginNode) {
				this.highlightBuilding(arrowTargetNode)
				const color = ColorConverter.getNumber(state.appSettings.mapColors.outgoingEdge)
				this.setCurveColor(curve, color)
			} else {
				this.highlightBuilding(arrowOriginNode)
				const color = ColorConverter.getNumber(state.appSettings.mapColors.incomingEdge)
				this.setCurveColor(curve, color)
			}
		}
	}

	addEdgePreview(nodes: Node[], edges: Edge[]) {
		if (nodes) {
			this.map = this.getNodesAsMap(nodes)
		}

		for (const edge of edges) {
			const originNode = this.map.get(edge.fromNodeName)
			const targetNode = this.map.get(edge.toNodeName)
			if (originNode && targetNode && edge.visible !== EdgeVisibility.none && edge.visible) {
				const curveScale = 100 * this.storeService.getState().appSettings.edgeHeight
				const curve = this.createCurve(originNode, targetNode, curveScale)
				this.previewMode(curve, edge.visible)
			}
		}
	}

	scale() {
		const { scaling } = this.storeService.getState().appSettings
		for (const arrow of this.arrows) {
			arrow.scale.x = scaling.x
			arrow.scale.y = scaling.y
			arrow.scale.z = scaling.z
		}
	}

	private isEdgeApplicableForBuilding(codeMapBuilding: CodeMapBuilding) {
		return this.storeService.getState().dynamicSettings.edgeMetric !== "None" && !codeMapBuilding.node.flat
	}

	private showEdgesOfBuildings(hoveredbuilding?: CodeMapBuilding) {
		const edges = this.storeService.getState().fileSettings.edges
		const buildings: Map<string, Node> = new Map()
		const selectedBuilding = this.threeSceneService.getSelectedBuilding()

		if (selectedBuilding) {
			const { node } = selectedBuilding
			buildings.set(node.path, node)
		}
		if (hoveredbuilding) {
			const { node } = hoveredbuilding
			buildings.set(node.path, node)
		}
		if (buildings.size > 0) {
			this.buildPairingEdges(buildings, edges)
		} else {
			this.addEdgePreview(
				null,
				edges.filter(x => x.visible !== EdgeVisibility.none)
			)
		}
	}

	private buildPairingEdges(node: Map<string, Node>, edges: Edge[]) {
		for (const edge of edges) {
			const originNode = this.map.get(edge.fromNodeName)
			const targetNode = this.map.get(edge.toNodeName)
			if (originNode && targetNode && node.has(originNode.path)) {
				this.addArrow(targetNode, originNode, true)
			} else if (originNode && targetNode && node.has(targetNode.path)) {
				this.addArrow(targetNode, originNode, false)
			}
		}
		this.threeSceneService.highlightBuildings()
	}

	private createCurve(arrowOriginNode: Node, arrowTargetNode: Node, curveScale) {
		const bezierPoint2 = arrowOriginNode.outgoingEdgePoint.clone()
		const bezierPoint3 = arrowTargetNode.incomingEdgePoint.clone()
		const arrowHeight = Math.max(bezierPoint2.y + arrowTargetNode.height, bezierPoint3.y + 1) + curveScale
		bezierPoint2.setY(arrowHeight)
		bezierPoint3.setY(arrowHeight)
		return new CubicBezierCurve3(arrowOriginNode.outgoingEdgePoint, bezierPoint2, bezierPoint3, arrowTargetNode.incomingEdgePoint)
	}

	private highlightBuilding(node: Node) {
		const building = this.threeSceneService.getMapMesh().getMeshDescription().getBuildingByPath(node.path)
		this.threeSceneService.addBuildingToHighlightingList(building)
	}

	private setCurveColor(bezier: CubicBezierCurve3, color: number, bezierPoints = 50) {
		const points = bezier.getPoints(bezierPoints)
		const curveObject = this.buildLine(points, color)
		curveObject.add(this.buildArrow(points))
		this.threeSceneService.edgeArrows.add(curveObject)
		this.arrows.push(curveObject)
	}

	private previewMode(curve, edgeVisibility: EdgeVisibility) {
		if (edgeVisibility === EdgeVisibility.both || edgeVisibility === EdgeVisibility.from) {
			const outgoingArrow: Object3D = this.makeArrowFromBezier(curve, false)
			this.threeSceneService.edgeArrows.add(outgoingArrow)
			this.arrows.push(outgoingArrow)
		}

		if (edgeVisibility === EdgeVisibility.both || edgeVisibility === EdgeVisibility.to) {
			const incomingArrow: Object3D = this.makeArrowFromBezier(curve, true)
			this.threeSceneService.edgeArrows.add(incomingArrow)
			this.arrows.push(incomingArrow)
		}
	}

	private getNodesAsMap(nodes: Node[]) {
		const map = new Map<string, Node>()
		nodes.forEach(node => map.set(node.path, node))
		return map
	}

	private makeArrowFromBezier(bezier: CubicBezierCurve3, incoming: boolean, bezierPoints = 50) {
		const points = bezier.getPoints(bezierPoints)
		let pointsPreviews: Vector3[]
		let arrowColor: string

		if (incoming) {
			pointsPreviews = points.slice(bezierPoints + 1 - this.VERTICES_PER_LINE)
			arrowColor = this.storeService.getState().appSettings.mapColors.incomingEdge
		} else {
			pointsPreviews = points
				.reverse()
				.slice(bezierPoints + 1 - this.VERTICES_PER_LINE)
				.reverse()
			arrowColor = this.storeService.getState().appSettings.mapColors.outgoingEdge
		}

		return this.buildEdge(pointsPreviews, ColorConverter.getNumber(arrowColor))
	}

	private buildEdge(points: Vector3[], color: number): Object3D {
		const curveObject = this.buildLine(points, color)
		curveObject.add(this.buildArrow(points))

		return curveObject
	}

	private buildLine(points: Vector3[], color = 0) {
		const geometry = new BufferGeometry()
		geometry.setFromPoints(points)

		const material = new LineBasicMaterial({ color, linewidth: 1 })
		return new Line(geometry, material)
	}

	private buildArrow(points: Vector3[], ARROW_COLOR = 0, headLength = 10, headWidth = 10) {
		const direction = points[points.length - 1]
			.clone()
			.sub(points[points.length - 2].clone())
			.normalize()

		const origin = points[points.length - 1].clone()
		if (direction.y < 0) {
			origin.y += headLength + 1
		}
		return new ArrowHelper(direction, origin, headLength + 1, ARROW_COLOR, headLength, headWidth)
	}
}
