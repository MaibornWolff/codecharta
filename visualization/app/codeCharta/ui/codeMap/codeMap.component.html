<div id="codeMap" ng-hide="$ctrl._viewModel.isLoadingFile">
	<view-cube-component ng-class="{ sideBarVisible: $ctrl._viewModel.isSideBarVisible }"></view-cube-component>
	<attribute-side-bar-component></attribute-side-bar-component>
</div>
