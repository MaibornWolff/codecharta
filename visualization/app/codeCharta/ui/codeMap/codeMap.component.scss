code-map-component {
	view-cube-component {
		position: absolute;
		z-index: 11;
		right: 0px;
		top: 10px;

		&.sideBarVisible {
			right: 350px;
		}
	}
}
