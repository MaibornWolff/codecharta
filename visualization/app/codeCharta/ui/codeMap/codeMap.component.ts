import { CodeMapBuilding } from "./rendering/codeMapBuilding"
import { ThreeViewerService } from "./threeViewer/threeViewerService"
import { BuildingRightClickedEventSubscriber, CodeMapMouseEventService } from "./codeMap.mouseEvent.service"

import "./codeMap.component.scss"

import { IRootScopeService, ITimeoutService } from "angular"
import { NodeContextMenuController } from "../nodeContextMenu/nodeContextMenu.component"
import { LoadingGifComponentSubscriber, LoadingGifService } from "../loadingGif/loadingGif.service"

export class CodeMapController implements BuildingRightClickedEventSubscriber, LoadingGifComponentSubscriber {
	private _viewModel: {
		isLoadingFile: boolean
	} = {
		isLoadingFile: true
	}

	/* @ngInject */
	constructor(
		private $rootScope: IRootScopeService,
		private $timeout: ITimeoutService,
		private $element: Element,
		private threeViewerService: ThreeViewerService,
		private codeMapMouseEventService: CodeMapMouseEventService
	) {
		CodeMapMouseEventService.subscribeToBuildingRightClickedEvents(this.$rootScope, this)
		LoadingGifService.subscribe(this.$rootScope, this)
	}

	public $postLink() {
		this.threeViewerService.init(this.$element[0].children[0])
		this.threeViewerService.animate()
		this.codeMapMouseEventService.start()
	}

	public onBuildingRightClicked(building: CodeMapBuilding, x: number, y: number) {
		NodeContextMenuController.broadcastHideEvent(this.$rootScope)
		if (building) {
			const nodeType = building.node.isLeaf ? "File" : "Folder"
			NodeContextMenuController.broadcastShowEvent(this.$rootScope, building.node.path, nodeType, x, y)
		}
	}

	public onLoadingFileStatusChanged(isLoadingFile: boolean) {
		this._viewModel.isLoadingFile = isLoadingFile
		this.synchronizeAngularTwoWayBinding()
	}

	public onLoadingMapStatusChanged(isLoadingMap: boolean) {}

	private synchronizeAngularTwoWayBinding() {
		this.$timeout(() => {})
	}
}

export const codeMapComponent = {
	selector: "codeMapComponent",
	template: require("./codeMap.component.html"),
	controller: CodeMapController
}
