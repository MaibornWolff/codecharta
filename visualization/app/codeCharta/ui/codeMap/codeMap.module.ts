import angular from "angular"

import "./threeViewer/threeViewer.module"
import "../../state/state.module"

import { codeMapComponent } from "./codeMap.component"
import { CodeMapMouseEventService } from "./codeMap.mouseEvent.service"
import { CodeMapRenderService } from "./codeMap.render.service"
import { CodeMapActionsService } from "./codeMap.actions.service"
import { CodeMapLabelService } from "./codeMap.label.service"
import { CodeMapArrowService } from "./codeMap.arrow.service"
import { CodeMapPreRenderService } from "./codeMap.preRender.service"
import camelCase from "lodash.camelcase"

angular
	.module("app.codeCharta.ui.codeMap", ["app.codeCharta.state", "app.codeCharta", "app.codeCharta.ui.codeMap.threeViewer"])
	.component(codeMapComponent.selector, codeMapComponent)
	.service(camelCase(CodeMapRenderService.name), CodeMapRenderService)
	.service(camelCase(CodeMapPreRenderService.name), CodeMapPreRenderService)
	.service(camelCase(CodeMapMouseEventService.name), CodeMapMouseEventService)
	.service(camelCase(CodeMapActionsService.name), CodeMapActionsService)
	.service(camelCase(CodeMapLabelService.name), CodeMapLabelService)
	.service(camelCase(CodeMapArrowService.name), CodeMapArrowService)
