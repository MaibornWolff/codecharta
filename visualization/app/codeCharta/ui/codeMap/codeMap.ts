import angular from "angular";

import "./threeViewer/threeViewer";
import "../../core/core.module";

import {codeMapComponent} from "./codeMap.component";
import {CodeMapMouseEventService} from "./codeMap.mouseEvent.service";
import {CodeMapRenderService} from "./codeMap.render.service";
import {CodeMapActionsService} from "./codeMap.actions.service";
import {CodeMapUtilService} from "./codeMap.util.service";
import {CodeMapLabelService} from "./codeMap.label.service";
import {CodeMapArrowService} from "./codeMap.arrow.service";

angular.module("app.codeCharta.ui.codeMap",["app.codeCharta.ui.codeMap.threeViewer", "app.codeCharta.core"])
.component(
    codeMapComponent.selector,
    codeMapComponent
).service(
    CodeMapRenderService.SELECTOR,
    CodeMapRenderService
).service(
    CodeMapMouseEventService.SELECTOR,
    CodeMapMouseEventService
).service(
    CodeMapActionsService.SELECTOR,
    CodeMapActionsService
).service(
    CodeMapUtilService.SELECTOR,
    CodeMapUtilService
).service(
    CodeMapLabelService.SELECTOR,
    CodeMapLabelService
).service(
    CodeMapArrowService.SELECTOR,
    CodeMapArrowService
);