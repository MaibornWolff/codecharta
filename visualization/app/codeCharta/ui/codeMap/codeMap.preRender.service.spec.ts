import "./codeMap.module"
import "../../codeCharta.module"
import { CodeMapRenderService } from "./codeMap.render.service"
import { CCFile, CodeMapNode, FileMeta, FileState, MetricData } from "../../codeCharta.model"
import { IRootScopeService } from "angular"
import { getService, instantiateModule } from "../../../../mocks/ng.mockhelper"
import { FileStateService } from "../../state/fileState.service"
import { MetricService } from "../../state/metric.service"
import { TEST_FILE_WITH_PATHS, METRIC_DATA, withMockedEventMethods, FILE_STATES, STATE } from "../../util/dataMocks"
import { CodeMapPreRenderService } from "./codeMap.preRender.service"
import { EdgeMetricDataService } from "../../state/edgeMetricData.service"
import { NodeDecorator } from "../../util/nodeDecorator"
import _ from "lodash"
import { StoreService } from "../../state/store.service"
import { ScalingService } from "../../state/store/appSettings/scaling/scaling.service"
import { setDynamicSettings } from "../../state/store/dynamicSettings/dynamicSettings.actions"
import { ScalingActions } from "../../state/store/appSettings/scaling/scaling.actions"
import { Vector3 } from "three"
import { IsLoadingMapActions } from "../../state/store/appSettings/isLoadingMap/isLoadingMap.actions"
import { BlacklistActions } from "../../state/store/fileSettings/blacklist/blacklist.actions"

describe("codeMapPreRenderService", () => {
	let codeMapPreRenderService: CodeMapPreRenderService
	let $rootScope: IRootScopeService
	let storeService: StoreService
	let fileStateService: FileStateService
	let metricService: MetricService
	let codeMapRenderService: CodeMapRenderService
	let edgeMetricDataService: EdgeMetricDataService

	let file: CCFile
	let fileMeta: FileMeta
	let map: CodeMapNode
	let fileStates: FileState[]
	let metricData: MetricData[]

	beforeEach(() => {
		restartSystem()
		rebuildService()
		withMockedEventMethods($rootScope)
		withMockedFileStateService()
		withMockedCodeMapRenderService()
		withMockedMetricService()
		withUnifiedMapAndFileMeta()
		storeService.dispatch(setDynamicSettings(STATE.dynamicSettings))
	})

	afterEach(() => {
		jest.resetAllMocks()
	})

	function restartSystem() {
		instantiateModule("app.codeCharta.ui.codeMap")

		$rootScope = getService<IRootScopeService>("$rootScope")
		storeService = getService<StoreService>("storeService")
		fileStateService = getService<FileStateService>("fileStateService")
		metricService = getService<MetricService>("metricService")
		codeMapRenderService = getService<CodeMapRenderService>("codeMapRenderService")
		edgeMetricDataService = getService<EdgeMetricDataService>("edgeMetricDataService")

		file = _.cloneDeep(TEST_FILE_WITH_PATHS)
		fileMeta = _.cloneDeep(FILE_STATES[0].file.fileMeta)
		map = _.cloneDeep(TEST_FILE_WITH_PATHS.map)
		map.children[1].children = _.slice(map.children[1].children, 0, 2)
		fileStates = _.cloneDeep(FILE_STATES)
		NodeDecorator.preDecorateFile(fileStates[0].file)
		metricData = _.cloneDeep(METRIC_DATA)
	}

	function rebuildService() {
		codeMapPreRenderService = new CodeMapPreRenderService(
			$rootScope,
			storeService,
			fileStateService,
			metricService,
			codeMapRenderService,
			edgeMetricDataService
		)
	}

	function withMockedCodeMapRenderService() {
		codeMapRenderService = codeMapPreRenderService["codeMapRenderService"] = jest.fn().mockReturnValue({
			render: jest.fn(),
			scaleMap: jest.fn()
		})()
	}

	function withMockedFileStateService() {
		fileStateService = codeMapPreRenderService["fileStateService"] = jest.fn().mockReturnValue({
			getFileStates: jest.fn().mockReturnValue(fileStates),
			fileStatesAvailable: jest.fn().mockReturnValue(true),
			isDeltaState: jest.fn().mockReturnValue(false)
		})()
	}

	function withMockedMetricService() {
		metricService = codeMapPreRenderService["metricService"] = jest.fn().mockReturnValue({
			getMetricData: jest.fn().mockReturnValue(metricData),
			isMetricAvailable: jest.fn().mockReturnValue(true)
		})()
	}

	function withUnifiedMapAndFileMeta() {
		codeMapPreRenderService["unifiedMap"] = map
		codeMapPreRenderService["unifiedFileMeta"] = fileMeta
	}

	describe("constructor", () => {
		beforeEach(() => {
			MetricService.subscribe = jest.fn()
			StoreService.subscribe = jest.fn()
			ScalingService.subscribe = jest.fn()
		})

		it("should call subscribe for MetricService", () => {
			rebuildService()

			expect(MetricService.subscribe).toHaveBeenCalledWith($rootScope, codeMapPreRenderService)
		})

		it("should call subscribe for StoreService", () => {
			rebuildService()

			expect(StoreService.subscribe).toHaveBeenCalledWith($rootScope, codeMapPreRenderService)
		})

		it("should call subscribe for ScalingService", () => {
			rebuildService()

			expect(ScalingService.subscribe).toHaveBeenCalledWith($rootScope, codeMapPreRenderService)
		})
	})

	describe("getRenderMap", () => {
		it("should return unifiedMap", () => {
			codeMapPreRenderService["unifiedMap"] = file.map

			const result = codeMapPreRenderService.getRenderMap()

			expect(result).toEqual(file.map)
		})
	})

	describe("onStoreChanged", () => {
		it("should call codeMapRenderService.render", done => {
			codeMapPreRenderService.onStoreChanged("SOME_ACTION")

			setTimeout(() => {
				expect(codeMapRenderService.render).toHaveBeenCalled()
				expect(storeService.getState().appSettings.isLoadingFile).toBeFalsy()
				expect(storeService.getState().appSettings.isLoadingMap).toBeFalsy()
				done()
			}, CodeMapPreRenderService["DEBOUNCE_TIME"])
		})

		it("should not call codeMapRenderService.render for scaling actions", () => {
			codeMapPreRenderService.onStoreChanged(ScalingActions.SET_SCALING)

			expect(codeMapRenderService.render).not.toHaveBeenCalled()
		})

		it("should not call codeMapRenderService.render for scaling actions", () => {
			codeMapPreRenderService.onStoreChanged(IsLoadingMapActions.SET_IS_LOADING_MAP)

			expect(codeMapRenderService.render).not.toHaveBeenCalled()
		})

		it("should show and stop the loadingMapGif", done => {
			codeMapPreRenderService["showLoadingMapGif"] = jest.fn()

			codeMapPreRenderService.onStoreChanged(BlacklistActions.SET_BLACKLIST)

			setTimeout(() => {
				expect(storeService.getState().appSettings.isLoadingMap).toBeFalsy()
				done()
			}, CodeMapPreRenderService["DEBOUNCE_TIME"])
		})
	})

	describe("onScalingChanged", () => {
		it("should call codeMapRenderService.render", () => {
			codeMapPreRenderService.onScalingChanged(new Vector3(1, 2, 3))

			expect(codeMapRenderService.scaleMap).toHaveBeenCalled()
		})

		it("should show and stop the loadingMapGif", () => {
			codeMapPreRenderService["showLoadingMapGif"] = jest.fn()

			codeMapPreRenderService.onScalingChanged(new Vector3(1, 2, 3))

			expect(codeMapPreRenderService["showLoadingMapGif"]).toHaveBeenCalled()
			expect(storeService.getState().appSettings.isLoadingMap).toBeFalsy()
		})
	})

	describe("onMetricDataAdded", () => {
		it("should decorate and set a new render map", () => {
			codeMapPreRenderService.onMetricDataAdded(metricData)

			expect(codeMapPreRenderService.getRenderMap()).toMatchSnapshot()
		})
	})
})
