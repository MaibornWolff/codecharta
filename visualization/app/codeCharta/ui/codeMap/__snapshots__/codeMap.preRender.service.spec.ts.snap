// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codeMapPreRenderService onMetricDataAdded should decorate and set a new render map 1`] = `
Object {
  "attributes": Object {
    "mcc": 111,
    "rloc": 200,
    "unary": 1,
  },
  "children": Array [
    Object {
      "attributes": Object {
        "functions": 10,
        "mcc": 1,
        "rloc": 100,
        "unary": 1,
      },
      "edgeAttributes": Object {},
      "link": "http://www.google.de",
      "name": "big leaf",
      "path": "/root/big leaf",
      "type": "File",
      "visible": true,
    },
    Object {
      "attributes": Object {
        "mcc": 110,
        "rloc": 100,
        "unary": 1,
      },
      "children": Array [
        Object {
          "attributes": Object {
            "functions": 100,
            "mcc": 100,
            "rloc": 30,
            "unary": 1,
          },
          "edgeAttributes": Object {},
          "name": "small leaf",
          "path": "/root/Parent Leaf/small leaf",
          "type": "File",
          "visible": true,
        },
        Object {
          "attributes": Object {
            "functions": 1000,
            "mcc": 10,
            "rloc": 70,
            "unary": 1,
          },
          "edgeAttributes": Object {},
          "name": "other small leaf",
          "path": "/root/Parent Leaf/other small leaf",
          "type": "File",
          "visible": true,
        },
      ],
      "edgeAttributes": Object {},
      "name": "Parent Leaf",
      "path": "/root/Parent Leaf",
      "type": "Folder",
      "visible": true,
    },
  ],
  "edgeAttributes": Object {},
  "name": "root",
  "path": "/root",
  "type": "Folder",
  "visible": true,
}
`;
